### PR TITLE
Modify rule S5270: Improve rule precision

### PR DESCRIPTION
**Pull request by michael-jabbour-sonarsource**
Related to CPP-3733 & SonarSource/sonar-cpp#2284.

### CODE REVIEW

File: rules/S5270/cfamily/rule.adoc [hunk 1]
```diff
@@ -30,17 +31,53 @@ void f() {
...
 
-  std::string myString = "foo";
-  v(myString.c_str()); // Compliant
+[source,cpp]
+----
+#include <numeric>
+#include <vector>
+
+// The variadic function is used as a catch-all net to terminate recursion
+std::size_t elementsCount(...) { return 1u; }  // Compliant
```
Comment (amelie-renard-sonarsource) [inline rules/S5270/cfamily/rule.adoc:63]: The "Compliant" should be on the line where `elementsCount` is called, as in the above examples.

File: rules/S5270/cfamily/rule.adoc [hunk 2]
```diff
@@ -30,17 +31,53 @@ void f() {
...
+    return count + elementsCount(element);
+  });
 }
 ----
 
+== See
+
+* MISRA C:2004, 16.1 - Functions shall not be defined with a variable number of arguments.
+* https://wiki.sei.cmu.edu/confluence/x/5ns-BQ[CERT, DCL50-CPP.] - Do not define a C-style variadic function
+* https://github.com/isocpp/CppCoreGuidelines/blob/036324/CppCoreGuidelines.md#-es34-dont-define-a-c-style-variadic-function[{cpp} Core Guidelines ES.34] Don't define a (C-style) variadic function
```
[the remark this answers is not in the document]
Comment (michael-jabbour-sonarsource) [reply rules/S5270/cfamily/rule.adoc:78]: From my perspective, the linked guidelines are only "relevant" to the rule, particularly where the rule suggests avoiding variadic functions for class types. My intention behind adding the links was to provide "further related reading material" around the topic.

I would find adding those tags to be a bit misleading here since the rule is less strict than the linked guidelines. I am also open to removing the whole "See" section if it is confusing...

File: rules/S5270/cfamily/rule.adoc [hunk 3]
```diff
@@ -1,27 +1,28 @@
...
 
 [source,cpp]
 ----
-class A {
-  char* toStr();
-};
-void v(...);
+#include <chrono>
+#include <string>
+#include <string_view>
```
[the remark this answers is not in the document]
Comment (michael-jabbour-sonarsource) [reply rules/S5270/cfamily/rule.adoc]: Thanks for pointing that out. Includes removed from all examples.

File: rules/S5270/cfamily/rule.adoc [hunk 4]
```diff
@@ -1,27 +1,28 @@
 Variadic arguments allow a function to accept any number of arguments (in this rule, we are not talking about variadic templates, but about functions with ellipses). But these arguments have to respect some criteria to be handled properly.
 
+The standard imposes some requirements on the types that can be passed as variadic arguments, and those requirements vary according to the C++ standard version in use. This rule detects any violations of these requirements since they can trigger undefined behavior.
```
[the remark this answers is not in the document]
Comment (michael-jabbour-sonarsource) [reply rules/S5270/cfamily/rule.adoc]: I originally mentioned the rules and the difference among standard versions, but then I removed them because the paragraph looked a bit hard to follow for me. I also think that those requirements are not the main point of the rule, since the rule forbids all class types anyway.

I added a minor change to make it clearer that we are talking about "class types" here, see 3d1a0bfa43a456a0393f927fd3b15754ac299858. Do you think that it is better to add another paragraph (or maybe something like a footnote) to describe these requirements?

[the remark this answers is not in the document]
Comment (amelie-renard-sonarsource) [reply rules/S5270/cfamily/rule.adoc]: Let's discuss it when you have time

File: rules/S5270/cfamily/rule.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (michael-jabbour-sonarsource) [reply rules/S5270/cfamily/rule.adoc:74]: Done in 16ce9dd9c7a8de1c39a17fb78c116b5b0579d149.

File: rules/S5270/cfamily/rule.adoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (michael-jabbour-sonarsource) [reply rules/S5270/cfamily/rule.adoc]: As discussed, I added a small explanation about the standard requirements in bb4a7ef18ba914b33cfb4acfe5fbc7fe7f03ec0a.

File: rules/S5270/cfamily/rule.adoc [hunk 5]
```diff
@@ -1,27 +1,30 @@
...
 
+The standard imposes some requirements on the class types that can be passed as variadic arguments, and those requirements vary according to the C++ standard version in use:
 
-This rules reports an issue if the type of the argument: 
+* Before C++11, the standard only allows POD types to be used as variadic arguments.
+* In C++11, the rules are relaxed such that any class type with an eligible non-trivial copy constructor, an eligible non-trivial move constructor, or a non-trivial destructor can be used in variadic arguments.
 
-* is a non trivially copyable, movable or deletable type: there is no guarantee that it will work.
-* is a class type that is trivially copyable, movable and deletable: in this case, we consider that the user intention was probably not to directly pass it as an argument but to call a method on it (``++c_str()++`` for example)
+Most importantly, the rule detects any violations of these requirements since they can trigger undefined behavior.
```
Comment (amelie-renard-sonarsource) [inline rules/S5270/cfamily/rule.adoc:8]:
```suggestion
The rule detects any violations of these requirements since they can trigger undefined behavior.
```

**Review by amelie-renard-sonarsource — CHANGES_REQUESTED** [review]
Minor suggestion otherwise LGTM!
I find the description clearer with the good amount of info

**Review by amelie-renard-sonarsource — APPROVED** [review]
